cardano-api/src/Cardano/Api/Fees.hs

```diff
@@ -1019,10 +1031,92 @@ makeTransactionBodyAutoBalance eraInMode systemstart history pparams
           txFee  = TxFeeExplicit explicitTxFees fee,
           txOuts = accountForNoChange
                      (TxOut changeaddr balance TxOutDatumNone ReferenceScriptNone)
-                     (txOuts txbodycontent)
+                     (txOuts txbodycontent),
+          txReturnCollateral = retColl,
+          txTotalCollateral = reqCol
         }
     return (BalancedTxBody txbody3 (TxOut changeaddr balance TxOutDatumNone ReferenceScriptNone) fee)
  where
+   -- Essentially we check for the existence of collateral inputs. If they exist we
+   -- create a fictious collateral return output. Why? Because we need to put dummy values
+   -- to get a fee estimate (i.e we overestimate the fee.)
+   maybeDummyTotalCollAndCollReturnOutput
+     :: TxBodyContent BuildTx era -> AddressInEra era -> (TxReturnCollateral CtxTx era, TxTotalCollateral era)
+   maybeDummyTotalCollAndCollReturnOutput TxBodyContent{txInsCollateral, txReturnCollateral, txTotalCollateral} cAddr =
+     case txInsCollateral of
+       TxInsCollateralNone -> (TxReturnCollateralNone, TxTotalCollateralNone)
+       TxInsCollateral{} ->
+         case totalAndReturnCollateralSupportedInEra era' of
+           Nothing -> (TxReturnCollateralNone, TxTotalCollateralNone)
+           Just retColSup ->
+             let dummyRetCol = TxReturnCollateral
+                                 retColSup
+                                 (TxOut cAddr (lovelaceToTxOutValue $ Lovelace (2^(64 :: Integer)) - 1)
+                                 TxOutDatumNone ReferenceScriptNone)
+                 dummyTotCol = TxTotalCollateral retColSup (Lovelace (2^(32 :: Integer) - 1))
+             in case (txReturnCollateral, txTotalCollateral) of
+                  (rc@TxReturnCollateral{}, tc@TxTotalCollateral{}) -> (rc, tc)
+                  (rc@TxReturnCollateral{},TxTotalCollateralNone) -> (rc, dummyTotCol)
+                  (TxReturnCollateralNone,tc@TxTotalCollateral{}) -> (dummyRetCol, tc)
+                  (TxReturnCollateralNone, TxTotalCollateralNone) -> (dummyRetCol, dummyTotCol)
+
+   calcReturnAndTotalCollateral
```

Comment:
Please add docs and links to any supporting documentation.

Reply:
Updated comments in the code